### PR TITLE
Greatly optimise selective masking vector exports

### LINKED ISSUE #50734: Unable to print or export
### What is the bug or the crash?

When attempting to print or export my map, QGIS works for a long time (193 seconds in the recorded case), then crashes abruptly. No warnings, no errors, just blip and the entire program disappears. If an output file is created, it has 0 size.

I have successfully exported exactly once to a 96 dpi PDF in Letter size, but I need 600 dpi and much larger paper size as there is more data to be added to this map yet.

I have tried exporting to PDF with multiple different options, exporting to PNG, and printing directly. All of the above fail in the same way.

I have tried reducing the canvas and layer sizes drastically by creating new layers with only the polygons etc. that I need, and removing the country-scale original layers.

This is not a hardware issue. QGIS doesn't ***even*** use a full 2 CPU threads of the 40 available, and only uses 324 MB of the available 96 GB RAM. Yes, the QGIS max thread count is set to 40 . . . as if that mattered.

The only crash output from DebugView is the following:
[14804] Warning: QObject::killTimer: Timers cannot be stopped from another thread
[14804] Warning: QObject::~QObject: Timers cannot be stopped from another thread

Here's the project: [temp QGIS project.zip](https://github.com/qgis/QGIS/files/9891365/temp.QGIS.project.zip)

### Steps to reproduce the issue

Try printing or exporting the attached project.

### Versions

Have tried QGIS versions 3.26 and 3.28.
Windows 10 build 19044.

### Supported QGIS version

- [X] I'm running a supported QGIS version according to the roadmap.

### New profile

- [X] I tried with a new QGIS profile

### Additional context

Clean install, no QGIS plugins installed.

### LINKED ISSUE #54788: Export produces huge PDF file when label mask are enabled in 3.28.11 compared to 3.22.16
### What is the bug or the crash?

With the same project file, opened in 3.22.16 and 3.28.11, I am getting a huge difference in file size when exporting to PDF.
1 Mb vs 100 Mb.

Only the case when I activate the label mask.

### Steps to reproduce the issue

Add a mask on the label of layer A.
Activate it on the mask tab in the properties of layer B.

If activated, the PDF export is very long and produces huge PDF in 3.28.11, not in 3.22.16.

### Versions

Version de QGIS
3.28.11-Firenze
Révision du code
b02458a196
Version de Qt
5.15.3
Version de Python
3.9.5
Version de GDAL/OGR
3.7.2
Version de Proj
9.3.0
Version de la base de données du registre EPSG
v10.094 (2023-08-08)
Version de GEOS
3.12.0-CAPI-1.18.0
Version de SQLite
3.41.1
Version de PDAL
2.5.5
Version du client PostgreSQL
unknown
Version de SpatiaLite
5.1.0
Version de QWT
6.1.6
Version de QScintilla2
2.13.4
Version de l'OS
Windows 10 Version 2009

Extensions Python actives
active_fire
0.3
CustomSVG
0.1
DigitizingTools
1.5.0
french_locator_filter
1.0.4
GroupStats
2.1.5
ImportPhotos
3.0.5
latlontools
3.6.15
lizmap
3.14.1
lockzoom
3.0.3
mask
1.10.2
MemoryLayerSaver
4.0.4
menu_from_project
v2.0.8
nominatim_locator_filter
0.2.4
OSMDownloader
1.0.3
PluginLoadTimes
3.0.1
processing_saga_nextgen
0.0.7
project_report
1.2
qfieldsync
3.1.5
qgis2web
3.16.0
qgis_preview_mode_plugin
0.1.0
QNEAT3
1.0.4
quick_map_services
0.19.11.1
refFunctions
1.6
SpreadsheetLayers
2.1.0
StreetView
3.2
db_manager
0.1.20
grassprovider
2.12.99
MetaSearch
0.3.6
processing
2.12.99
sagaprovider
2.12.99

### Supported QGIS version

- [X] I'm running a supported QGIS version according to [the roadmap](https://www.qgis.org/en/site/getinvolved/development/roadmap.html#release-schedule).

### New profile

- [X] I tried with a new [QGIS profile](https://docs.qgis.org/latest/en/docs/user_manual/introduction/qgis_configuration.html#working-with-user-profiles)

### Additional context

With a new profile, QGIS crashed during the export.

Test project downloadable : https://acloud.zaclys.com/index.php/s/aK2cZGTjYt4gHA4

Layer "Centre d'incendie et de secours" has a label mask
Layers Commune and Département have the mask activated with the label mask of "Centre d'incendie et de secours".

### PR DESCRIPTION
**Pull request by nyalldawson**
Optimise the logic used when the new geometry backend for selective masking is in effect:

Whenever its SAFE, instead of calculating an "entire map" clipping path and then applying this for every feature being rendered,
we now defer the calculation of the clipping path until we are rendering individual features. Then, we create a clipping path
which contains ONLY the mask paths which are within the area being drawn over.

This avoids having the entire map clipping path being used for EVERY feature being rendered, which results in huge PDF/SVG
exports when masks are in effect, and instead results in clipping paths which are confined just to a sensible area
around each rendered feature.

In some complex test projects this reduces the PDF export size by a factor of 0.01, without any loss of quality or regressions in export time!! (Eg 60Mb -> 700kb). It also results in PDFs/SVGs which open much quicker in viewers and editors, and don't grind their operation to a halt.

Fixes https://github.com/qgis/QGIS/issues/50734
Fixes https://github.com/qgis/QGIS/issues/54788

### CODE REVIEW

**Comment by nyalldawson** [comment]
@troopa81 here's the next optimisation for selective masking exports. I consider this one the "holy grail", which squashes the last of the remaining big limitations of the vectorised selective masking.

File: src/core/symbology/qgssymbollayer.cpp [hunk 1]
```diff
@@ -1000,20 +1005,55 @@ void QgsSymbolLayer::prepareMasks( const QgsSymbolRenderContext &context )
...
+  bool res = false;
   if ( !mClipPath.isEmpty() )
   {
     context.painter()->save();
     context.painter()->setClipPath( mClipPath, Qt::IntersectClip );
+    res = true;
+  }
+  else if ( rect.isValid() )
+  {
+    // find just the clip geometries within the area the symbol layer will be drawn over
```
Comment (troopa81) [inline src/core/symbology/qgssymbollayer.cpp:1019]: It looks like pretty much the same piece of code of prepareMasks, can we factor it a bit in a method ?

**Review by troopa81 — COMMENTED** [review]
I made a first quick review, I'll try to finish it in a few hours.

Having a tailored mask for each feature is interesting, but I think it brings a lot of complexity to the code (specially in QgsSymbol::renderFeature). 

Like I said before, I'm wondering if we could not have improved Qt by sharing/referening clippingPath when writing PDF. This way, we would have had one global mask (maybe built and simplified using the geometry paint device) applied to all features. In the end, the crash appears when serializing, not painting.

Therefore, the more feature you have, the bigger the file will be because you multiply masks, while with a global mask your will size stay constant.

File: src/core/symbology/qgssymbollayerutils.h [hunk 2]
```diff
@@ -929,6 +929,18 @@ class CORE_EXPORT QgsSymbolLayerUtils
...
+     * Returns a list of the symbol layer clip geometries to be used for the symbol layer with the specified
+     * ID.
+     *
+     * The \a bounds argument specifies the target bounds (in painter coordinates) for matching geometries. Only mask
+     * geometries which intersect \a bounds will be returned. If \a bounds is a null QRectF then all clip geometries
+     * for the symbol layer will be returned.
+     *
+     * \since QGIS 3.38
+     */
+    static QVector< QgsGeometry > collectSymbolLayerClipGeometries( const QgsRenderContext &context, const QString &symbolLayerId, const QRectF &bounds );
```
[the remark this answers is not in the document]
Comment (nyalldawson) [reply src/core/symbology/qgssymbollayerutils.h:942]: QgsRenderContext already is quite a large API -- I'd rather keep utility functions out of there as much as possible, and leave it focused on just getter/setter properties.

File: src/core/symbology/qgssymbollayerutils.cpp [hunk 3]
```diff
@@ -5418,6 +5418,26 @@ void QgsSymbolLayerUtils::resetSymbolLayerIds( QgsSymbolLayer *symbolLayer )
...
+
+  if ( bounds.isNull() )
+    return clipGeometries;
+
+  const QgsRectangle boundsRect = QgsRectangle( bounds );
+
+  clipGeometries.erase(
+    std::remove_if( clipGeometries.begin(), clipGeometries.end(), [&boundsRect]( const QgsGeometry & geometry )
+  {
+    return !geometry.boundingBoxIntersects( boundsRect );
```
[the remark this answers is not in the document]
Comment (nyalldawson) [reply src/core/symbology/qgssymbollayerutils.cpp:5435]: @troopa81 

> Did you consider returning an intersection so we lower even more the size of the mask? or is it overkill and the intersection would be too much time consuming?

When geometries are added by the geometry paint device they are all split into singlepart geometry objects. That means that each geometry here will correspond to a single character (or symbol component), so I don't think we'll gain much advantage from doing an actual intersection here vs the cost of the intersection calculation itself.

**Comment by nyalldawson** [comment]
@troopa81 
> I made a first quick review, I'll try to finish it in a few hours.

Thanks!

> Like I said before, I'm wondering if we could not have improved Qt by sharing/referening clippingPath when writing PDF. This way, we would have had one global mask (maybe built and simplified using the geometry paint device) applied to all features. 

> Therefore, the more feature you have, the bigger the file will be because you multiply masks, while with a global mask your will size stay constant.

That sounds good, and it would definitely be nice to see this happen in some future Qt release! :+1:  I've now added a render context/layout flag which forces the "entire map" clipping paths to always be used. It's opt in, but if Qt does improve the situation then we'll easily be able to switch back by just enabling this flag. (At some stage I'd love to make some advanced dialog for configuring layout export presets, and then we could even expose this setting to users who want fine control over this, and all the other mask simplification parameters too).

Pragmatically though, there's no related Qt improvements in current Qt 6 versions and use of local clip masks fixes the bulk of the issues with exporting complex layouts, so we should make this default behaviour for now.

> In the end, the crash appears when serializing, not painting.

It's caused during serializing the painter **state**, because the clip path is too complex for the structures Qt internally is using to do this. So we'd actually need two upstream improvements -- one would be the "write clip path once when exporting to SVG/PDF" improvement, and the other would be "adapt QPainter state storage to handle complex clip paths". 

On the other hand, with the changes here I can confirm that use of local clip paths fixes all the crashing projects from the linked tickets with current Qt stable versions.

File: src/core/symbology/qgssymbol.h [hunk 4]
```diff
@@ -329,9 +329,10 @@ class CORE_EXPORT QgsSymbol
      * \param fields fields for features to be rendered (usually the associated
      * vector layer's fields). Required for correct calculation of data defined
      * overrides.
+     * \param isSubSymbol Should be set to TRUE when the symbol is being used as a subsymbol of a QgsSymbolLayer (since QGIS 3.38)
      * \see stopRender()
      */
-    void startRender( QgsRenderContext &context, const QgsFields &fields = QgsFields() );
+    void startRender( QgsRenderContext &context, const QgsFields &fields = QgsFields(), bool isSubSymbol = false );
```
[the remark this answers is not in the document]
Comment (nyalldawson) [reply src/core/symbology/qgssymbol.h:335]: @troopa81 indeed -- and it turns out there's a natural place for this already, in the symbols Qgis::SymbolRenderHint flags!

I've implemented this now.

File: src/core/symbology/qgssymbollayer.cpp (hunk 1 above)
Comment (nyalldawson) [reply src/core/symbology/qgssymbollayer.cpp:1019]: Ok, done

**Comment by troopa81** [comment]
Is it normal ?
layout_export_mask

**Comment by nyalldawson** [comment]
@troopa81 

> Is it normal

Yes, it's relating to that qt issue regarding paths bleeding out of the map frame's clip area. I'm fairly certain I can fix that now by doing the intersection of map item clip path and selective masking clip path via the geometry approach, but I'll do that in a different PR. And if it works I'll regenerate all the reference images so that none of them have the bleed anymore.

File: src/core/symbology/qgsarrowsymbollayer.cpp [hunk 5]
```diff
@@ -215,7 +215,9 @@ void QgsArrowSymbolLayer::startRender( QgsSymbolRenderContext &context )
   mComputedHeadType = headType();
   mComputedArrowType = arrowType();
 
-  mSymbol->startRender( context.renderContext() );
+  mSymbol->setRenderHints( mSymbol->renderHints() | Qgis::SymbolRenderHint::IsSymbolLayerSubSymbol );
```
Comment (troopa81) [inline src/core/symbology/qgsarrowsymbollayer.cpp:218]: Wouldn't be better to do it when adding the symbol to its parent symbol layer so the flag could be reused outside of the rendering pipeline?

File: src/core/symbology/qgssymbol.cpp [hunk 6]
```diff
@@ -1642,6 +1663,23 @@ void QgsSymbol::renderFeature( const QgsFeature &feature, QgsRenderContext &cont
     if ( needsExpressionContext )
       mSymbolRenderContext->expressionContextScope()->addVariable( QgsExpressionContextScope::StaticVariable( QStringLiteral( "symbol_layer_index" ), symbolLayerIndex + 1, true ) );
 
+    // if this symbol layer has associated clip masks, we need to render it to a QPicture first so that we can
+    // determine the actual rendered bounds of the symbol. We'll then use that to retrieve the clip masks we need
+    // to apply when painting the symbol via this QPicture.
+    const bool hasClipGeometries = !context.testFlag( Qgis::RenderContextFlag::AlwaysUseGlobalMasks )
```
Comment (troopa81) [inline src/core/symbology/qgssymbol.cpp:1669]: Shall we not test `!mRenderHints.testFlag( Qgis::SymbolRenderHint::IsSymbolLayerSubSymbol` ? Isn't that the opposite of the condition above?

Comment (nyalldawson) [reply src/core/symbology/qgssymbol.cpp]: Right, fixed

File: src/core/symbology/qgsarrowsymbollayer.cpp (hunk 5 above)
Comment (nyalldawson) [reply src/core/symbology/qgsarrowsymbollayer.cpp:218]: Let's defer that till it's actually needed. Turns out it's not so trivial to do, as the render hints aren't cloned (and I don't think they should be, either). The logic ends up much more complex as a result.

File: src/core/symbology/qgssymbol.cpp [hunk 7]
```diff
@@ -1633,6 +1654,9 @@ void QgsSymbol::renderFeature( const QgsFeature &feature, QgsRenderContext &cont
   if ( needsExpressionContext )
     mSymbolRenderContext->expressionContextScope()->addVariable( QgsExpressionContextScope::StaticVariable( QStringLiteral( "symbol_layer_count" ), mLayers.count(), true ) );
 
+  const bool maskGeometriesDisabledForSymbol = context.testFlag( Qgis::RenderContextFlag::AlwaysUseGlobalMasks )
+                                               && !mRenderHints.testFlag( Qgis::SymbolRenderHint::IsSymbolLayerSubSymbol );
```
Comment (nyalldawson) [inline src/core/symbology/qgssymbol.cpp:1658]:
```suggestion
    && !mRenderHints.testFlag( Qgis::SymbolRenderHint::IsSymbolLayerSubSymbol );
```

File: src/core/symbology/qgssymbol.cpp [hunk 8]
```diff
@@ -1633,6 +1654,9 @@ void QgsSymbol::renderFeature( const QgsFeature &feature, QgsRenderContext &cont
   if ( needsExpressionContext )
     mSymbolRenderContext->expressionContextScope()->addVariable( QgsExpressionContextScope::StaticVariable( QStringLiteral( "symbol_layer_count" ), mLayers.count(), true ) );
 
+  const bool maskGeometriesDisabledForSymbol = context.testFlag( Qgis::RenderContextFlag::AlwaysUseGlobalMasks )
+    && !mRenderHints.testFlag( Qgis::SymbolRenderHint::IsSymbolLayerSubSymbol );
```
Comment (nyalldawson) [inline src/core/symbology/qgssymbol.cpp:1658]:
```suggestion
      && !mRenderHints.testFlag( Qgis::SymbolRenderHint::IsSymbolLayerSubSymbol );
```

**Pull request merged**